src/main/java/io/cdap/plugin/servicenow/sink/service/ServiceNowSinkAPIRequestImpl.java

```diff
@@ -0,0 +1,251 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.servicenow.sink.service;
+
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.WaitStrategies;
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.plugin.servicenow.apiclient.ServiceNowTableAPIClientImpl;
+import io.cdap.plugin.servicenow.apiclient.ServiceNowTableAPIRequestBuilder;
+import io.cdap.plugin.servicenow.restapi.RestAPIResponse;
+import io.cdap.plugin.servicenow.sink.ServiceNowSinkConfig;
+import io.cdap.plugin.servicenow.sink.model.RestRequest;
+import io.cdap.plugin.servicenow.sink.model.ServiceNowBatchRequest;
+import io.cdap.plugin.servicenow.util.ServiceNowConstants;
+import org.apache.http.Header;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.message.BasicHeader;
+import org.apache.oltu.oauth2.common.exception.OAuthProblemException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import javax.ws.rs.core.MediaType;
+
+/**
+ * Implementation class for ServiceNow Batch Rest API.
+ */
+public class ServiceNowSinkAPIRequestImpl {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowSinkAPIRequestImpl.class);
+
+  private static Integer counter = 1;
+  private static Integer batchRequestIdCounter = 1;
+  private final ServiceNowSinkConfig config;
+  private final ServiceNowTableAPIClientImpl restApi;
+  private final Gson gson = new Gson();
+  private final JsonParser jsonParser = new JsonParser();
+  private boolean isCreated;
+
+
+  public ServiceNowSinkAPIRequestImpl(ServiceNowSinkConfig conf) {
+    this.config = conf;
+    restApi = new ServiceNowTableAPIClientImpl(config);
+  }
+
+  public RestRequest getRestRequest(JsonObject jsonObject) {
+    JsonElement jsonElement = jsonObject.get(ServiceNowConstants.SYS_ID);
+    String sysId = null;
+
+    if (jsonElement == null) {
```

Comment:
can we merge both if statements and what impact it will have if we don't remove sys_id while using insert operation.

Reply:
Some of the tables do not support the insertion of records with `sys_id` and throw 403 Permission Denied error. Also, it is not a required field as it is created by the ServiceNow application after the successful insertion of the record. 

Merging both the if statements are not recommended.
If I simplify like this : 

```
if (jsonElement == null && config.getOperation().equals(ServiceNowConstants.UPDATE_OPERATION)) {
      throw new RuntimeException("No sys_id found in the record to be updated");
    } else if (config.getOperation().equals(ServiceNowConstants.INSERT_OPERATION)) {
      jsonObject.remove(ServiceNowConstants.SYS_ID);
    }
sysId = jsonElement.getAsString();
```

then the second case fails when trying to remove the `sys_id`
CASE I :   `sys_id` is present with INSERT operation Result: PASS 
CASE II :  `sys_id` is not present with INSERT operation Result: FAIL 
CASE III : `sys_id` is present with UPDATE operation Result: PASS
CASE IV : `sys_id` is not present with UPDATE operation Result: PASS

Reply:
jsonElement != null && config.getOperation().equals(ServiceNowConstants.INSERT_OPERATION)
add this in second else if

Reply:
This statement `sysId = jsonElement.getAsString();` would give NullPointerException if `jsonElement` is `null` and if we move this statement inside the **else if** block, then we won't get the `sys_id` value for _UPDATE_ operation